src/meta/src/stream/stream_graph.rs

```diff
@@ -184,359 +175,399 @@ impl StreamActorBuilder {
+        self.rewrite_inner(&self.nodes, 0)
+    }
+
+    fn rewrite_inner(&self, stream_node: &StreamNode, depth: usize) -> MetaResult<StreamNode> {
+        match stream_node.get_node_body()? {
+            // Leaf node `Exchange`.
+            NodeBody::Exchange(exchange) => {
+                // The exchange node should always be the bottom of the plan node. If we find one
+                // when the depth is 0, it means that the plan node is not well-formed.
+                if depth == 0 {
+                    bail!(
+                        "there should be no ExchangeNode on the top of the plan node: {:#?}",
+                        stream_node
+                    )
                 }
-            } else if self.upstreams.values().any(|u| u.same_worker_node) {
-                Some(ColocatedActorId {
-                    id: self
-                        .upstreams
-                        .values()
-                        .filter(|x| x.same_worker_node)
-                        .exactly_one()
-                        .unwrap()
-                        .actors
-                        .as_global_ids()
-                        .into_iter()
-                        .exactly_one()
-                        .unwrap(),
+                assert!(!stream_node.get_fields().is_empty());
+                assert!(stream_node.input.is_empty());
+
+                // Index the upstreams by the an internal edge ID.
+                let upstreams = &self.upstreams[&EdgeId::Internal {
+                    link_id: stream_node.get_operator_id(),
+                }];
+
+                Ok(StreamNode {
+                    node_body: Some(NodeBody::Merge(MergeNode {
+                        upstream_actor_id: upstreams.actors.as_global_ids(),
+                        upstream_fragment_id: upstreams.fragment_id.as_global_id(),
+                        upstream_dispatcher_type: exchange.get_strategy()?.r#type,
+                        fields: stream_node.get_fields().clone(),
+                    })),
+                    identity: "MergeExecutor".to_string(),
+                    ..stream_node.clone()
                 })
-            } else {
-                None
-            },
-            vnode_bitmap: None,
-            // To be filled by `StreamGraphBuilder::build`
-            mview_definition: "".to_owned(),
+            }
+
+            // "Leaf" node `Chain`.
+            NodeBody::Chain(chain_node) => {
+                let input = stream_node.get_input();
+                assert_eq!(input.len(), 2);
+
+                let merge_node = &input[0];
+                assert_matches!(merge_node.node_body, Some(NodeBody::Merge(_)));
+                let batch_plan_node = &input[1];
+                assert_matches!(batch_plan_node.node_body, Some(NodeBody::BatchPlan(_)));
+
+                // Index the upstreams by the an external edge ID.
+                let upstreams = &self.upstreams[&EdgeId::UpstreamExternal {
```

Comment:
Note that this edge is external, that is, not included in the `FragmentGraph` but the `CompleteGraph`. To distinguish between these two kinds of edges, we use different edge ID variants here.